src/script/interpreter.h

```diff
@@ -114,32 +120,101 @@ enum
     // Making OP_CODESEPARATOR and FindAndDelete fail any non-segwit scripts
     //
     SCRIPT_VERIFY_CONST_SCRIPTCODE = (1U << 16),
+
+    // Taproot/Tapscript validation (BIPs 341 & 342)
+    //
+    SCRIPT_VERIFY_TAPROOT = (1U << 17),
+
+    // Making unknown Taproot leaf versions non-standard
+    //
+    SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION = (1U << 18),
+
+    // Making the use of (unknown) annexes non-standard (currently no annexes are known)
+    //
+    SCRIPT_VERIFY_DISCOURAGE_UNKNOWN_ANNEX = (1U << 19),
+
+    // Making unknown OP_SUCCESS non-standard
+    SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS = (1U << 20),
+
+    // Making unknown public key versions (in BIP 342 scripts) non-standard
+    SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_PUBKEYTYPE = (1U << 21),
 };
 
 bool CheckSignatureEncoding(const std::vector<unsigned char> &vchSig, unsigned int flags, ScriptError* serror);
 
 struct PrecomputedTransactionData
 {
+    //! Single-SHA256 versions
+    uint256 m_prevouts_hash, m_sequences_hash, m_outputs_hash, m_amounts_spent_hash;
+    bool m_amounts_spent_ready = false;
+
+    //! Double-SHA256 versions
     uint256 hashPrevouts, hashSequence, hashOutputs;
     bool ready = false;
+    std::vector<CTxOut> m_spent_outputs;
+
+    PrecomputedTransactionData() = default;
+
+    template <class T>
+    void Init(const T& tx, std::vector<CTxOut> spent_outputs);
 
     template <class T>
     explicit PrecomputedTransactionData(const T& tx);
 };
 
 enum class SigVersion
 {
-    BASE = 0,
-    WITNESS_V0 = 1,
+    BASE = 0,        //!< Bare scripts and P2SH redeemscripts; see BIP 16
+    WITNESS_V0 = 1,  //!< Witness v0 (P2WPKH and P2WSH); see BIP 141
+    TAPROOT = 2,     //!< Witness v1 with non-P2SH 32 byte program (Taproot), key path spending; see BIP 341
+    TAPSCRIPT = 3,   //!< Witness v1 with non-P2SH 32 byte program (Taproot), script path spending, leaf version 0xc0 (Tapscript); see BIP 342
+};
+
+struct ScriptExecutionData
```

Comment:
I'm generally not a fan of these Data interfaces which have all the bool x_init fields.

As an alternative, we could use an optional type wrapper.

However, optional types depend on boost until c++17, which we shouldn't introduce consensus dependencies on.

My proposal would be to copy the definition of c++17 optional into https://github.com/bitcoin/bitcoin/blob/99813a9745fe10a58bedd7a4cb721faf14f907a4/src/optional.h and drop the boost dependency, and then use this here. This can be done as a PR separate from Taproot so as not to further burden this PR.


Here, optionals express both the nothing set and not initialized state (hence dual-wrapped options) but perhaps we'd be OK in collapsing the wrapped ones to represent (not initialized or not set) OR value.

```c++
 struct ScriptExecutionData
 {
     //! The tapleaf hash.
     Option<uint256> m_tapleaf_hash;
     Option<Option<uint32_t>> m_codeseparator_pos;
     //! Hash of the annex data, if any
     Option<Option<uint256>> m_annex_hash;
     /** How much validation weight is left (decremented for every successful signature check). */
     Option<int64_t> m_validation_weight_left;
 };
```

This would seem, to me, to make the interface quite a bit safer and eliminates a bunch of initialization logic & prevents UB from cropping up.

It would also be, in my opinion, a larger project (perhaps worthwhile) to generally rethink this state object to be accessed through an interface that guarantees the fields are accessed correctly -- e.g., once a m_validation_weight_left is set it can only be read or decreased, and the function to decrease can set error if the condition isn't met on underflow. Maybe not worth that large of a refactor though, I think optionals would already be an improvement.

Reply:
I agree to an extent that all the _init variables are ugly, but I don't think that they're that bad if you see them as just runtime checks for invalid code. The same can be accomplished without them using ubsan/valgrind, in many cases.

About using Option... It's arguably the right tool for the job, but there is no singular "definition of c++17 optional"; there is the libstdc++ implementation, and the libc++ one, both of which may be relying on platform specific extensions (I haven't checked). Even ignoring that, the libstdc++ one is 1500 lines of code, which I think is really overkill for what we'd get. Since the Boost one can't be used in libconsensus, I think the current code is not too bad.

Reply:
I examined libstdc++ implementation for RISC-V architechture and it seems portable and independent from any physical architecture
https://github.com/riscv/riscv-gcc/commits/riscv-gcc-8.3.0/libstdc%2B%2B-v3

Reply:
@StEvUgnIn We can't just copy GPL code, unfortunately.

Reply:
@JeremyRubin Don't forget that the `std` namespace get special treatment from the compiler.
so copying code from `libstd` without checking it might actually contain undefined behavior for anyone outside of the `std` namespace.

Reply:
@JeremyRubin Don't forget that the `std` namespace get special treatment from the compiler.
so copying code from `libstd` without checking it might actually contain implementation defined behavior that won't necessarily work correctly with a different compiler(or even just by being outside of the std namespace).

Reply:
This seems to be more of a potential improvement for later.